### PR TITLE
[WIP/DNM] src/config: type-safe config parsing

### PR DESCRIPTION
**Pull request by andelf**
WIP.

POW 😆

### CODE REVIEW

File: src/config/mod.rs [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+    storage: StorageConfig,
+
+    #[serde(rename = "pd")]
+    #[serde(skip_serializing)]
+    pd_depercated: Option<PdConfig>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(default, rename_all = "kebab-case")]
+struct ServerConfig {
```
Comment (siddontang) [inline src/config/mod.rs]: put config to its own package, don't combine all into one place

File: src/config/mod.rs [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+    cluster_id: Option<u64>,
+
+    // TODO: IPv6 support via std::net::SocketAddr
+    addr: SocketAddrV4,
+
+    #[serde(deserialize_with = "types::deserialize_opt_addr")]
+    advertise_addr: Option<SocketAddrV4>,
+
+    #[serde(with = "types::addrs")]
+    pd_endpoints: Vec<SocketAddrV4>,
```
Comment (siddontang) [inline src/config/mod.rs]: V6 support?

File: src/config/mod.rs [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+    #[serde(rename = "pd")]
+    #[serde(skip_serializing)]
+    pd_depercated: Option<PdConfig>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(default, rename_all = "kebab-case")]
+struct ServerConfig {
+    #[serde(default, skip_serializing, skip_deserializing)]
+    cluster_id: Option<u64>,
```
Comment (siddontang) [inline src/config/mod.rs]: why Option?

File: src/config/mod.rs [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+    #[serde(rename = "store")]
+    #[serde(skip_serializing)]
+    store_deprecated: Option<PathBuf>,
+}
+
+impl Default for ServerConfig {
+    fn default() -> ServerConfig {
+        ServerConfig {
+            cluster_id: None,
+            addr: "127.0.0.1:20160".parse().unwrap(),
```
Comment (siddontang) [inline src/config/mod.rs]: define const values  for all the magic values

File: src/config/types/duration.rs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            .build()
+            .unwrap();
+
+        let caps = re.captures(value).unwrap();
+
+        let value = caps["value"].parse().map_err(|e| E::custom(format!("error parsing: {:?}", e)));
+        let unit = caps["unit"].to_lowercase();
+
+        match unit.as_ref() {
+            "" | "s" => value.map(Duration::from_secs),
```
Comment (siddontang) [inline src/config/types/duration.rs]: /cc @BusyJay 

The time unit is millisecond or second? how do you think?

**Comment by siddontang** [comment]
Please add test.

File: src/config/mod.rs (hunk 3 above)
Comment (andelf) [inline src/config/mod.rs]: should be removed

File: src/config/mod.rs (hunk 1 above)
Comment (andelf) [inline src/config/mod.rs]: reject

File: src/config/mod.rs (hunk 2 above)
Comment (andelf) [inline src/config/mod.rs]: TODO

File: src/config/mod.rs (hunk 4 above)
Comment (andelf) [inline src/config/mod.rs]: const only used once, no need to do so.
just gather them all in `impl Default`

File: src/config/types/duration.rs (hunk 5 above)
Comment (andelf) [inline src/config/types/duration.rs]: metric is the standard.

File: src/config/mod.rs (hunk 4 above)
Comment (siddontang) [inline src/config/mod.rs]: use const

**Pull request closed without merging**